### PR TITLE
Image Stream v2

### PR DESCRIPTION
**Pull request by schlan**
### Changes
- Completely reworked the existing ImageStream
  - Moved to a PopupWindow (`ImageStreamUI`) for hosting the UI and a Fragment (`ImageStream`) for handling permissions and media resolving 
  - The image stream aims to always exactly cover the keyboard.
- Optimize the sizing of images in the stream.
- Support for multi select
- Support for providing selected items on startup. There's also support for adding extra items onto the top of the list. 
- Integrated the new floating action menu
- Add support for rendering non image attachments. 
- Updated BelvedereUi
- Added ChatActivity to sample app. For testing and demonstrating how the image stream works.
- Change the Logger to be a static class

This is the initial import of the new UI. Tests, Javadoc and improvements are following. 

### Reviewers
@brendan-fahy @baz8080 @schlan @eepDev

### FYI
@KevinLambe

### CODE REVIEW

File: belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java [hunk 1]
```diff
@@ -19,31 +19,30 @@
...
 
-    private final static String LOG_TAG = "Belvedere";
+    final static String LOG_TAG = "Belvedere";
 
     @SuppressLint("StaticFieldLeak")
     private static Belvedere instance;
 
     private final Context context;
-    private final Logger log;
     private final boolean debug;
```
Comment (eepDev) [inline belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java]: Unused variable

File: belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java [hunk 2]
```diff
@@ -19,31 +19,30 @@
...
 
     private Storage storage;
     private IntentRegistry intentRegistry;
     private MediaSource mediaSource;
 
     Belvedere(Builder builder) {
         this.context = builder.context;
-        this.log = builder.logger;
+
+        builder.logger.setLoggable(true);
```
Comment (eepDev) [inline belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java]: Loggable always true

File: belvedere-core/src/main/java/zendesk/belvedere/Storage.java [hunk 3]
```diff
@@ -214,44 +208,44 @@ File getFile(Context context, String subDirectory, String fileName) {
...
-     *     If the directory not exist the method tries to create it.
+     * If the directory not exist the method tries to create it.
      * </p>
      *
-     * @param context A valid application {@link Context}
+     * @param context      A valid application {@link Context}
      * @param subDirectory The name of the sub directory
      * @return A {@link File} pointing to the directory.
      */
     @SuppressWarnings("ResultOfMethodCallIgnored")
```
Comment (eepDev) [inline belvedere-core/src/main/java/zendesk/belvedere/Storage.java]: Is this annotation needed?

File: belvedere/src/main/java/zendesk/belvedere/BelvedereUi.java [hunk 4]
```diff
@@ -14,6 +20,123 @@
...
+            return this;
+        }
+
+        public ImageStreamBuilder withDocumentIntent(String contentType, boolean allowMultiple) {
+            this.mediaIntents.add(Belvedere.from(context).document().allowMultiple(allowMultiple).contentType(contentType).build());
+            return this;
+        }
+
+        public ImageStreamBuilder withSelectedItems(List<MediaResult> mediaResults) {
+            this.selectedItems = mediaResults;
```
Comment (eepDev) [inline belvedere/src/main/java/zendesk/belvedere/BelvedereUi.java]: Should we assign or add mediaResults?

Assigning can cause a side effect when first calling method that adds intent to list (camera, document) and then another method that reassigns mediaIntents.
For example, when first adding camera intent and then media intents, the original camera intent will be gone 
```
imageStreamBuilder.withCameraIntent().withSelectedItems(mediaResults)
```

File: belvedere/src/main/java/zendesk/belvedere/BelvedereUi.java [hunk 5]
```diff
@@ -14,6 +20,123 @@
...
+        private List<MediaIntent> mediaIntents = new ArrayList<>();
+        private List<MediaResult> selectedItems = new ArrayList<>();
+        private List<MediaResult> extraItems = new ArrayList<>();
+
+        private ImageStreamBuilder(Context context){
+            this.context = context;
+        }
+
+        public ImageStreamBuilder withMediaIntents(List<MediaIntent> mediaIntents) {
+            this.mediaIntents = mediaIntents;
```
Comment (eepDev) [inline belvedere/src/main/java/zendesk/belvedere/BelvedereUi.java]: Should we assign or add mediaResults?

Assigning can cause a side effect when first calling method that adds intent to list (camera, document) and then another method that reassigns mediaIntents.
For example, when first adding camera intent and then media intents, the original camera intent will be gone
```
imageStreamBuilder.withCameraIntent().withSelectedItems(mediaResults)
```

File: belvedere/src/main/java/zendesk/belvedere/ImageStreamPresenter.java [hunk 6]
```diff
@@ -1,85 +1,38 @@
...
 
 class ImageStreamPresenter implements ImageStreamMvp.Presenter {
 
     private final ImageStreamMvp.Model model;
     private final ImageStreamMvp.View view;
 
-    ImageStreamPresenter(ImageStreamMvp.Model model, ImageStreamMvp.View view) {
+    private final ImageStreamDataSource imageStreamDataSource;
+
+    ImageStreamPresenter(ImageStreamMvp.Model model, ImageStreamMvp.View view, ImageStreamDataSource imageStreamDataSource) {
```
Comment (eepDev) [inline belvedere/src/main/java/zendesk/belvedere/ImageStreamPresenter.java]: What's the purpose of `imageStreamDataSource` in this class? I cannot find any usages of it, do we need to pass it in a constructor and later store it in a field?

File: belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java (hunk 1 above)
Comment (schlan) [inline belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java]: Fixed.

File: belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java (hunk 2 above)
Comment (schlan) [inline belvedere-core/src/main/java/zendesk/belvedere/Belvedere.java]: Fixed. 
The whole logger concept is still a bit odd, I'll look into fixing it a bit later.

File: belvedere/src/main/java/zendesk/belvedere/BelvedereUi.java (hunk 5 above)
Comment (schlan) [inline belvedere/src/main/java/zendesk/belvedere/BelvedereUi.java]: Changed to always add intents.

File: belvedere/src/main/java/zendesk/belvedere/ImageStreamPresenter.java (hunk 6 above)
Comment (schlan) [inline belvedere/src/main/java/zendesk/belvedere/ImageStreamPresenter.java]: 'historical reasons'. I forgot the remove it ... :S
It's gone.

**Pull request merged**